### ISSUE TITLE
Issues with rotate

### ISSUE THREAD

**Issue by cestes** [Delgan/loguru#365]
I'm aware I could just be misusing loguru, but have noticed some non-intuitive behavior around 'rotate'.

I'm running:
`Python 3.7.6 (default, Feb 26 2020, 20:54:15) 
[GCC 7.3.1 20180712 (Red Hat 7.3.1-6)] on linux`
and
`loguru                 0.5.1`

I have a web-app built in [bottle](http://bottlepy.org/) that uses the gunicorn webserver and has 6 worker processes.

In my app I have the following code: `logger.add("pods.log", rotation="sunday at 03:00")`

I think I'm running into problems when I have multiple users on the website simultaneously and gunicorn runs multithreaded copies of the website.  This is actually kind of unusual; the site is kind of sandbox and the user volume is really low... we have maybe 5 visitors per day.  I have a bunch of single-threaded cron jobs that also rotate their logs like this one does, and it works fine.  However this is the only multithreaded program that I have and It seems to create lots more log files than I ask for.  Here's what the last few weeks look like:
```
-rw-r--r-- 1 root     root     135829 Oct 31 17:55 pods.2020-10-26_04-57-05_964506.log
-rw-r--r-- 1 root     root       6893 Nov  2 20:11 pods.2020-11-01_18-20-42_710982.log
-rw-r--r-- 1 root     root       1384 Nov  2 03:23 pods.2020-11-02_03-20-20_265978.log
-rw-r--r-- 1 root     root       2900 Nov  2 07:36 pods.2020-11-02_03-42-34_356316.log
-rw-r--r-- 1 root     root       3303 Nov  2 20:16 pods.2020-11-02_07-38-17_428417.log
-rw-r--r-- 1 root     root      92259 Nov  7 20:07 pods.2020-11-02_08-14-48_165825.log
-rw-r--r-- 1 root     root       5989 Nov 10 21:55 pods.2020-11-09_07-38-13_902695.log
-rw-r--r-- 1 root     root       4503 Nov 10 00:55 pods.2020-11-09_08-36-40_989960.log
-rw-r--r-- 1 root     root       6215 Nov 11 03:45 pods.2020-11-09_17-55-30_725461.log
-rw-r--r-- 1 root     root       5524 Nov 11 03:42 pods.2020-11-09_18-03-11_396160.log
-rw-r--r-- 1 root     root       4538 Nov 11 06:05 pods.2020-11-10_01-11-39_544732.log
-rw-r--r-- 1 root     root        562 Nov 10 01:56 pods.2020-11-10_01-56-49_697171.log
-rw-r--r-- 1 root     root      25173 Nov 13 17:08 pods.log
```
We can see on Nov 11, it created a lot of logs... not sure why; it's not even a Sunday.

Are my expectations wrong or am I doing something wrong?

Regardless -- I love this tool!  Thanks for making it!

**Comment by Delgan** [comment]
Hi @cestes. You're not the first one encountering difficulties mixing `loguru` with `gunicorn`, unfortunately. :(

Typically, multiple log files instead of one indicates an issue with `multiprocessing`. For example, looking at this code:

```python
from loguru import logger
import multiprocessing
import time

def worker():
    while True:
        time.sleep(0.01)
        logger.info("Logging...")


logger.add("foo.log", rotation="200 KB")

if __name__ == "__main__":
    multiprocessing.set_start_method("spawn")
    processes = [multiprocessing.Process(target=worker) for _ in range(4)]

    for p in processes:
        p.start()

    for p in processes:
        p.join()
```
At first, all logs appear in `foo.log` and it seems to work perfectly. However, as soon as the rotation takes place, several new files are created in which the processes write separately. This happens because actually `logger.add()` is called each time a process is started, which means your program is effectively logging to multiple handlers and not to the same file.

I don't know if it's the same problem you face with `gunicorn`, but it could be. Do you call `logger.add()` in a `if __name__ == "__main__"` block (so that it's configured only once: in the main process)? Also, maybe the workers were on standby, and only created the file when they was first used, although it was not a Sunday.

There is also this article about configuring `loguru` with `guvicorn`, maybe it can help you: [Unify Python logging for a Gunicorn/Uvicorn/FastAPI application](https://pawamoy.github.io/posts/unify-logging-for-a-gunicorn-uvicorn-app/).

In any case, I will try to help you investigate this problem, do not hesitate to ask me. ;)

**Comment by cestes** [comment]
Thanks for the quick and thoughtful feedback!  My gut tells me this has to do with multiprocessing.  In my example above on 11-Nov at around 03:42, two guys on our team were testing and we had the somewhat unusual case of using the multiprocessing feature of bottle/gunicorn.  One other time I noticed it, I think we had a classroom of people playing with our app.

I don't call logger.add() from my `__main__` .  Here's my code:

```
def main():
    logger.add("pods.log", rotation="sunday at 03:00")
    logger.info("Program starting")

    try:
        parser = configparser.ConfigParser()
        parser.read('pods.ini')
    except:
        logger.critical("Error reading config file")

    logger.info("Starting web application")

    run(host     = parser.get('webserver','host'), 
        port     = parser.get('webserver','port'),
        debug    = parser.getboolean('webserver','debug'),
        certfile = parser.get('webserver','certfile'),
        keyfile  = parser.get('webserver','keyfile'),
        server   = parser.get('webserver','server'),
        timeout  = parser.get('webserver','timeout'),
        workers  = parser.getint('webserver','workers')
        )

    logger.info("Program stopping")

if __name__ == '__main__':
    main()
```
The `run()` starts bottle/gunicorn.

If I move `logger.add()` from `main()` to `__main__` do you think that would help?

**Comment by Delgan** [comment]
Well, you are indirectly calling `logger.add()` in the `if __name__ == '__main__'` block as it's part of your `main()` function. So, moving `logger.add()` around would not change anything.

I think the problem does come from multiprocessing, but not exactly for the reason I first mentioned.

```python
from loguru import logger
import multiprocessing
import time

def worker():
    while True:
        time.sleep(0.01)
        logger.info("Logging...")



if __name__ == "__main__":
    logger.add("foo.log", rotation="200 KB")

    multiprocessing.set_start_method("fork")
    processes = [multiprocessing.Process(target=worker) for _ in range(4)]

    for p in processes:
        p.start()

    for p in processes:
        p.join()
```

The new processes are probably created using the `"fork"` method. In the end, the result is the same: each child process ends up with an independent handler but logging to the same file. This means every handler will try to rotate the file separately, causing the duplicated files you're seeing.

What you need is a way to synchronize the worker so that they effectively use one common handler. That way, the rotation will be triggered only once, and then all processes will keep logging to the same shared file.

You should try to use `enqueue=True` while configuring the `logger`:

```python
logger.add("pods.log", enqueue=True, rotation="sunday at 03:00")
```


  I also advise you to read this paragraph in the documentation that may help you understand how this parameter works: [Compatibility with `multiprocessing` using `enqueue` argument](https://loguru.readthedocs.io/en/stable/resources/recipes.html#compatibility-with-multiprocessing-using-enqueue-argument).

**Comment by Delgan** [comment]
@cestes Did that fix your problem? Can I close this issue?

**Comment by kylinat2688** [comment]
> @cestes Did that fix your problem? Can I close this issue?

I also meet this problem, and I have used the parameter equeue=True,the difference is that I add log in gunicorn hood when_ready,here is my code:

`import log_util
proc_name = "picture_finger_api"
workers = 8
worker_class = 'gevent'
bind = "0.0.0.0:7410"


def on_starting(server):
    log_util.init_log('log/app.log', enqueue=True)`

the log_util add the log

**Comment by Delgan** [comment]
Hi @kylinat2688. 

I'm unable to reproduce the issue. Log files is properly rotated and compressed according to my tests.

Can you please provide a minimal reproducible example (feel free to open a new ticket)?

**Comment by kylinat2688** [comment]
Oh,I have found the bug, gunicorn worker time out when loguru is compressing the log. 
the log file is big, about  1.8G, gunicorn worker time out with 30s as default.  When I modify time out to 0 which means no timeout, It works right, or I remove the compress config, It's also work right. 
Reproduce code: git@github.com:kylinat2688/loguru_demo.git

But, I also have a question, is loguru compress log in worker process, not in the master process  when I config enqueue=True?

**Comment by Delgan** [comment]
@kylinat2688 Thanks for the link to the repository. When `enqueue=True`, the compression occurs in a separate worker thread started internally by the `logger`. Theoretically, it should not affect your processes.

**Comment by yejuan6** [comment]
@Delgan Hi Delgan， I also have the problem  when I use uvicorn to start multiprocess app.  Here is My code:

`

LoggerConfig("/mnt/data/log/app/console_server/", "console_server")

def main():
    uvicorn.run(
        app="console_server:app",
        host='0.0.0.0',
        port=settings.port,
        workers=16,
        lifespan="on",
    )

if __name__ == "__main__":
    multiprocessing.set_start_method("fork")
    main()
`
Here is LoggerConfig definition:
`
class LoggerConfig:
    def __init__(self, log_dir: str, service_name: str):
        ............. some init code .....
        self.json_log_file_prefix = os.path.join(self.log_dir, self.service_name)

        logger.remove()
        logger.add(
            self.json_log_file_prefix + ".info.log",
            level=settings.LOG_LEVEL,
            rotation="100 KB",
            retention=10,
            # compression="zip",
            enqueue=True,
        )
`

The rotated log like:
-rw-rw-r-- 1 zhang zhang  98K Dec 20 13:58 console_server.info.2024-12-20_13-58-12_185315.log
-rw-rw-r-- 1 zhang zhang 1.4K Dec 20 13:58 console_server.info.2024-12-20_13-58-12_191081.log
-rw-rw-r-- 1 zhang zhang 3.5K Dec 20 13:58 console_server.info.2024-12-20_13-58-12_206267.log
-rw-rw-r-- 1 zhang zhang 8.4K Dec 20 13:58 console_server.info.2024-12-20_13-58-12_236480.log
-rw-rw-r-- 1 zhang zhang  13K Dec 20 13:58 console_server.info.2024-12-20_13-58-12_375595.log
-rw-rw-r-- 1 zhang zhang  561 Dec 20 13:58 console_server.info.log

When console_server.info.log is 100 KB,  it created a lot of logs above.....
loguru    version         0.7.2

**Comment by Delgan** [comment]
Hi @yejuan6.

Unfortunately, it's currently not possible to use an unique file handler with non-synchronized workers such as the one spawned by Uvicorn. This is because the `logger` must be inherited by the child processes, but Uvicorn does not provide such API.

You would need to define a handler that synchronizes logs (using `filelock`, for example), or use a different file for each worker.

This is discussed here in the documentation: [Why are my log files sometimes duplicated or the content trimmed?](https://loguru.readthedocs.io/en/stable/resources/troubleshooting.html#why-are-my-log-files-sometimes-duplicated-or-the-content-trimmed).

**Comment by yejuan6** [comment]
> Hi @yejuan6.
> 
> Unfortunately, it's currently not possible to use an unique file handler with non-synchronized workers such as the one spawned by Uvicorn. This is because the `logger` must be inherited by the child processes, but Uvicorn does not provide such API.
> 
> You would need to define a handler that synchronizes logs (using `filelock`, for example), or use a different file for each worker.
> 
> This is discussed here in the documentation: [Why are my log files sometimes duplicated or the content trimmed?](https://loguru.readthedocs.io/en/stable/resources/troubleshooting.html#why-are-my-log-files-sometimes-duplicated-or-the-content-trimmed).

Thanks @Delgan for your reply !   

I found logger.add() function has a parameter "watch",  set watch=True, it means that in one process,  when  console_server.info.log reaches 100KB ,it will be rotated to console_server.info.2024-12-20_13-58-12_375595.log, next logger.info() can still output in console_server.info.log  but not in  console_server.info.2024-12-20_13-58-12_375595.log. 
So all processes will always output in console_server.info.log. 
Am I right?

**Comment by Delgan** [comment]
The `watch` argument is orthogonal to `multiprocessing` problems. It purposefully re-creates the file if it has been deleted or renamed, which can happen with Linux's `logrotate` tool, for example. This argument re-implements the [`logging.WatchedFileHandler`](https://docs.python.org/3/library/logging.handlers.html#watchedfilehandler) from the standard library.

I don't think it will help with the Uvicorn workers, as concurrency issues will remain due to lack of synchronization mechanism. Maybe it just hides the problem. After rotation, the handler always checks that the file to be created does not already exist, regardless of the value of `watch`.

In the next few days, I will try to think about and provide an advanced code recipe to handle `multiprocessing` with Uvicorn. This will probably be based on a socket server.

**Comment by Cre4T3Tiv3** [comment]
Hi @Delgan ,

Thanks for the update and for sharing your thoughts on handling multiprocessing challenges with Uvicorn.

I understand that the `watch` argument is orthogonal to this issue and that an effective solution would require a proper **synchronization mechanism** to avoid concurrency problems. Your mention of using a **socket server approach** to manage multiprocessing log rotation sounds very promising.

Would love to hear more about your planned approach! If there’s anything I can do to **help test, review, or refine the solution**, I’d be happy to contribute. 

Additionally, do you think it would make sense to **provide official guidance on integrating Loguru with external rotation tools like `logrotate`**, for users who prefer to delegate rotation to system-level utilities?

Looking forward to your insights! 🚀

**Comment by Delgan** [comment]
Hey @Cre4T3Tiv3. Thanks for your interest in helping Loguru! :+1: 

I recently updated the documentation recently with a code snippet that explains how to synchronize Loguru logs using a socket server: [Transmitting log messages across network, processes or Gunicorn workers](https://loguru.readthedocs.io/en/latest/resources/recipes.html#transmitting-log-messages-across-network-processes-or-gunicorn-workers).

This solution should be robust against workers-based frameworks such as Guvicorn.
If you have an application that would benefit from such a solution, please try it out and report any problems in the code or confusion in the documentation.

For the future, I'll come up with a more convenient API to integrate this directly into Loguru.

Regarding `logrotate`, if its use with Loguru is relevant, then an example would surely be welcome in the code snippets page.

**Comment by Cre4T3Tiv3** [comment]
Hi @Delgan,

Thanks for your detailed response and for pointing me to the updated documentation! The socket server approach looks like a well-structured solution for handling log synchronization across multiple worker processes. I will review the implementation and evaluate its behavior in a multiprocessing setup. If I come across any edge cases or areas where the documentation could be clarified, I will be sure to provide feedback.

I also appreciate your consideration of a more integrated API for this functionality—it sounds like a valuable enhancement.

Regarding logrotate, I will take some time to assess whether its integration with Loguru introduces any practical limitations or advantages. If it proves to be a viable approach that aligns well with common use cases, I would be happy to contribute an example or insights for the code snippets page.

**Comment by Cre4T3Tiv3** [comment]
Hi @Delgan 

Thanks for your detailed response and for pointing me to the updated documentation! The socket server approach looks like a well-structured solution for handling log synchronization across multiple worker processes. I will review the implementation and evaluate its behavior in a multiprocessing setup. If I come across any edge cases or areas where the documentation could be clarified, I will be sure to provide feedback.

I also appreciate your consideration of a more integrated API for this functionality—it sounds like a valuable enhancement.

Regarding logrotate, I will take some time to assess whether its integration with Loguru introduces any practical limitations or advantages. If it proves to be a viable approach that aligns well with common use cases, I would be happy to contribute an example or insights for the code snippets page.

**Comment by Cre4T3Tiv3** [comment]
Hi @Delgan,

Thanks for your detailed response and for pointing me to the updated documentation! The socket server approach looks like a well-structured solution for handling log synchronization across multiple worker processes. I will review the implementation and evaluate its behavior in a multiprocessing setup. If I come across any edge cases or areas where the documentation could be clarified, I will be sure to provide feedback.

I also appreciate your consideration of a more integrated API for this functionality—it sounds like a valuable enhancement.

Regarding logrotate, I will take some time to assess whether its integration with Loguru introduces any practical limitations or advantages. If it proves to be a viable approach that aligns well with common use cases, I would be happy to contribute an example or insights for the code snippets page.
